crates/ty_python_semantic/src/types/function.rs

```diff
@@ -0,0 +1,983 @@
+        Self::new(
+            db,
+            self.last_definition(db),
+            Some(inherited_generic_context),
+        )
+    }
+
+    fn update_last_definition(
+        self,
+        db: &'db dyn Db,
+        f: impl FnOnce(OverloadLiteral<'db>) -> OverloadLiteral<'db>,
+    ) -> Self {
+        Self::new(
+            db,
+            f(self.last_definition(db)),
+            self.inherited_generic_context(db),
+        )
+    }
+
+    fn name(self, db: &'db dyn Db) -> &'db ast::name::Name {
+        // All of the overloads of a function literal should have the same name.
+        self.last_definition(db).name(db)
+    }
+
+    fn known(self, db: &'db dyn Db) -> Option<KnownFunction> {
+        // Whether a function is known is based on its name (and its containing module's name), so
+        // all overloads should be known (or not) equivalently.
+        self.last_definition(db).known(db)
+    }
+
+    fn has_known_decorator(self, db: &dyn Db, decorator: FunctionDecorators) -> bool {
+        self.iter_overloads_and_implementation(db)
+            .any(|overload| overload.decorators(db).contains(decorator))
+    }
+
+    fn definition(self, db: &'db dyn Db) -> Definition<'db> {
+        self.last_definition(db).definition(db)
+    }
+
+    fn parameter_span(
+        self,
+        db: &'db dyn Db,
+        parameter_index: Option<usize>,
+    ) -> Option<(Span, Span)> {
+        self.last_definition(db).parameter_span(db, parameter_index)
+    }
+
+    fn spans(self, db: &'db dyn Db) -> Option<FunctionSpans> {
+        self.last_definition(db).spans(db)
+    }
+
+    #[salsa::tracked(returns(ref))]
+    fn overloads_and_implementation(
+        self,
+        db: &'db dyn Db,
+    ) -> (Box<[OverloadLiteral<'db>]>, Option<OverloadLiteral<'db>>) {
+        let self_overload = self.last_definition(db);
+        let mut current = self_overload;
+        let mut overloads = vec![];
+
+        while let Some(previous) = current.previous_overload(db) {
+            let overload = previous.last_definition(db);
+            overloads.push(overload);
+            current = overload;
+        }
+
+        // Overloads are inserted in reverse order, from bottom to top.
+        overloads.reverse();
+
+        let implementation = if self_overload.is_overload(db) {
+            overloads.push(self_overload);
+            None
+        } else {
+            Some(self_overload)
+        };
+
+        (overloads.into_boxed_slice(), implementation)
+    }
+
+    fn iter_overloads_and_implementation(
+        self,
+        db: &'db dyn Db,
+    ) -> impl Iterator<Item = OverloadLiteral<'db>> + 'db {
+        let (implementation, overloads) = self.overloads_and_implementation(db);
+        overloads.iter().chain(implementation).copied()
+    }
+
+    /// Typed externally-visible signature for this function.
+    ///
+    /// This is the signature as seen by external callers, possibly modified by decorators and/or
+    /// overloaded.
```

Comment:
This doc comment I think also deserves the same warning about not calling it cross-file. (I guess the fact that it's private should help here.) It's important that all cross-file access goes through the Salsa-tracked `FunctionType::signature`.

Reply:
Added.  I also added a similar comment to `OverloadLiteral::signature` above, which is crate-public, since it is called by the `display()` implementation, and when rendering diagnostics (which I think are both okay places to add the cross-file dependency?)